### ISSUE TITLE
Error while calling web service operation: queryMore 

### ISSUE THREAD

**Issue by xp224px** [forcedotcom/dataloader#576]
We have a job scheduled in Jenkins to call DataLoader CLI to extract data daily. Been having this issue every now and then.

FYI 
sfdc.connectionTimeoutSecs =270

2022-11-17 04:26:32,542 INFO  [xxxPolicyExtract] progress.NihilistProgressAdapter setSubTask (NihilistProgressAdapter.java:78) - Processed 586,000 of 589,103 total records. Rate: 1,183,000 records per hour. Estimated time to complete: 0 minutes and 9 seconds.  There are 586,000 successes and 0 errors.
2022-11-17 04:26:50,576 ERROR [xxxPolicyExtract] client.PartnerClient runOperation (PartnerClient.java:359) - Error while calling web service operation: queryMore, error was: 
com.sforce.soap.partner.fault.InvalidQueryLocatorFault: null

So basically 586,000 of 589,103 extracted and failed towards very end of this job.

see the log file attached
[Invalid Query Locator error.txt](https://github.com/forcedotcom/dataloader/files/10027483/Invalid.Query.Locator.error.txt)

**Comment by ashitsalesforce** [comment]
Hi @xp224px , 

A few questions about your setup:
- Which version of Data Loader are you using?
- Are you using Bulk or Partner API for the query? 
- What is your batch size?

**Comment by xp224px** [comment]
Which version of Data Loader are you using? 
- DataLoader version v53.0.3

Are you using Bulk or Partner API for the query? did not set bulk api as true.

- sfdc.bulkApiCheckStatusInterval=5000
- sfdc.bulkApiSerialMode=false
- sfdc.bulkApiZipContent=false
- sfdc.useBulkApi=false

What is your batch size?
-sfdc.extractionRequestSize=500
-dataAccess.readBatchSize=200
-dataAccess.writeBatchSize=500
-sfdc.loadBatchSize=200

**Comment by ashitsalesforce** [comment]
Thanks @xp224px. Can you try by setting "sfdc.useBulkApi=true" and "sfdc.readBatchSize=2000"?

**Comment by xp224px** [comment]
Hi @ashitsalesforce  thanks. will modify them and test it for a week to see the results
Could you please explain the rationale behind this ?  
What if bulk api cannot be set as true due to limitation of bulk api (truncation, insert null), any alternative solutions if thats the case ?

**Comment by ashitsalesforce** [comment]
Hi @xp224px , it is likely that the query cursor expires on the server side while data loader is making "query more" due to the large result size when using partner API. "query more" is done on the server side when using Bulk API, which avoids the cursor expiry. 

I do not understand your point about limitations of bulk api such as truncation, insert null that are not in partner api. Can you elaborate?

**Comment by xp224px** [comment]
@ashitsalesforce 
isn't when 
sfdc.useBulkAPi=false
then 
sfdc.truncateFields is NOT available ?

**Comment by ashitsalesforce** [comment]
Hi @xp224px , "sfdc.truncateFields=true" sets the request header "[AllowFieldTruncation](https://developer.salesforce.com/docs/atlas.en-us.apexref.meta/apexref/apex_methods_system_database_dmloptions.htm#apex_Database_DmlOptions_allowFieldTruncation)", which truncates a string field value during an insert operation based on the documentation. Given that yours is a query operation, this property is not applicable.

**Comment by xp224px** [comment]
@ashitsalesforce 
Thank you. Have not seen "[Error while calling web service operation: queryMore]" for a few days now while doing testing with your suggestions.

However, this error showed up
Could you let me know the cause and how could we avoid this ?

[Error Log 23_Nov.txt](https://github.com/forcedotcom/dataloader/files/10071949/Error.Log.23_Nov.txt)

**Comment by ashitsalesforce** [comment]
Hi @xp224px ,

Does this error occurs each time you run data loader or does it occur intermittently? Also, does it occur after extracting (and writing to the csv file on your desktop) specific number of rows or does the number of extracted rows differ each time?

**Comment by xp224px** [comment]
@ashitsalesforce  see files attached. it looks like it occurred intermittently. Also it doesn't look like the specific number of rows.  I also did 2 manual built in Jenkins invoking the job. 1 failed (also attached, same error) and 1 successful.  

On top of it, I ran DataLoader GUI version with & without bulk API enabled. there was no error occurred during extraction.

could you explain this and any solutions in terms of avoiding this ?

[Premature Error 23_Nov_2022.txt](https://github.com/forcedotcom/dataloader/files/10080759/Premature.Error.23_Nov_2022.txt)
[Premature Error 24_Nov_2022.txt](https://github.com/forcedotcom/dataloader/files/10080760/Premature.Error.24_Nov_2022.txt)
[Premature Error 24_Nov_2022_Manual_Built.txt](https://github.com/forcedotcom/dataloader/files/10080761/Premature.Error.24_Nov_2022_Manual_Built.txt)

**Comment by ashitsalesforce** [comment]
Hi @xp224px , 

Thanks for providing the additional context. Based on your observations and information provided by other java developers [here](https://stackoverflow.com/questions/72228406/why-premature-eof-error-has-occur-while-reading-csv-file-using-springboot-sched), it is highly likely that the connection between data loader and Salesforce server where it fetches data from occasionally terminates due to a glitch in the network. Chances of this happen increase if the connection is longer lasting, which can occur if the query results are large and they are retrieved slowly. 

While receiving query results data loader keeps the server connection open while processing and storing each row in the csv file. This results in having the connection open for a long time and increasing the chances of a network disruption resulting in premature termination of the connection. To mitigate this issue, you can set the property `loader.bufferUnprocessedBulkQueryResults=true`. Doing so results in data loader first downloading query results to a temporary file without any processing, closing the server connection, and finally processing the downloaded results and placing them in the csv file, thereby minimizing the amount of time during which the server connection is open. This fix was developed to address another related issue - issue #170 .

The temporary file where data loader stores server results when `loader.bufferUnprocessedBulkQueryResults=true` is located in the temporary folder. For Windows, the temporary folder is `%USER%\AppData\Local\Temp`. For Mac, the temporary folder is `/tmp`. The temporary file's prefix is `sdl` and its suffix is `.csv`. Make sure that the temporary folder has sufficient disk space to store the full extraction.

**Comment by xp224px** [comment]
@ashitsalesforce  thank you! will conduct some tests in upcoming weeks. will let you know how it goes

**Comment by ashitsalesforce** [comment]
Hi @xp224px , I am closing the issue assuming that it is resolved.
